### PR TITLE
Prod/Sandbox Endpoints change

### PR DESCRIPTION
**Pull request by Krishanx92**

### CODE REVIEW

File: adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go [hunk 1]
```diff
@@ -110,18 +110,30 @@ func (r *API) validateAPI() error {
...
+	if !(r.Spec.Production != nil && r.Spec.Production[0].HTTPRouteRefs != nil && len(r.Spec.Production[0].HTTPRouteRefs) > 0) && !(r.Spec.Sandbox != nil && r.Spec.Sandbox[0].HTTPRouteRefs != nil && len(r.Spec.Sandbox[0].HTTPRouteRefs) > 0) {
 		allErrs = append(allErrs, field.Required(field.NewPath("spec"),
 			"both API production and sandbox endpoint references cannot be empty"))
 	}
 
-	if isEmptyStringsInArray(r.Spec.ProdHTTPRouteRefs) {
-		allErrs = append(allErrs, field.Required(field.NewPath("spec").Child("prodHTTPRouteRef"),
+	var prodHTTPRoute1, sandHTTPRoute1 []string
+	if r.Spec.Production != nil {
+		prodHTTPRoute1 = r.Spec.Production[0].HTTPRouteRefs
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go:120]:
```suggestion
	if len(r.Spec.Production) > 0 {
	for prodEnv := range r.Spec.Production {
	isEmptyStringsInArray(prodEnv.HTTPRouteRefs) {
		allErrs = append(allErrs, field.Required(field.NewPath("spec").Child("production").Child("httpRouteRefs"),
			"API production endpoint reference cannot be empty"))
	}
	}
```

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 2]
```diff
@@ -205,9 +205,21 @@ func (apiReconciler *APIReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, nil
 	}
 
+	var prodHTTPRoute1, sandHTTPRoute1 []string
+	if apiDef.Spec.Production != nil {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:209]:
```suggestion
	if len(apiDef.Spec.Production) > 0 {
```

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 3]
```diff
@@ -205,9 +205,21 @@ func (apiReconciler *APIReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+		prodHTTPRoute1 = apiDef.Spec.Production[0].HTTPRouteRefs
+	} else {
+		prodHTTPRoute1 = []string{}
+	}
+	if apiDef.Spec.Sandbox != nil {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:214]:
```suggestion
	if len(apiDef.Spec.Sandbox) > 0 {
```

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 4]
```diff
@@ -900,22 +912,26 @@ func addIndexes(ctx context.Context, mgr manager.Manager) error {
...
 			api := rawObj.(*dpv1alpha1.API)
 			var httpRoutes []string
-			for _, ref := range api.Spec.ProdHTTPRouteRefs {
-				if ref != "" {
-					httpRoutes = append(httpRoutes,
-						types.NamespacedName{
-							Namespace: api.Namespace,
-							Name:      ref,
-						}.String())
+			if api.Spec.Production != nil {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:915]:
```suggestion
			if len(api.Spec.Production) > 0 {
```

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 5]
```diff
@@ -205,9 +205,21 @@ func (apiReconciler *APIReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, nil
 	}
 
+	var prodHTTPRoute1, sandHTTPRoute1 []string
+	if apiDef.Spec.Production != nil {
+		prodHTTPRoute1 = apiDef.Spec.Production[0].HTTPRouteRefs
+	} else {
+		prodHTTPRoute1 = []string{}
+	}
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:213]: i think we don't need this block

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 6]
```diff
@@ -205,9 +205,21 @@ func (apiReconciler *APIReconciler) Reconcile(ctx context.Context, req ctrl.Requ
...
+	if apiDef.Spec.Sandbox != nil {
+		sandHTTPRoute1 = apiDef.Spec.Sandbox[0].HTTPRouteRefs
+	} else {
+		sandHTTPRoute1 = []string{}
+	}
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:218]: I think we don't need this block

File: adapter/pkg/operator/controllers/dp/api_controller.go [hunk 7]
```diff
@@ -900,22 +912,26 @@ func addIndexes(ctx context.Context, mgr manager.Manager) error {
...
-			for _, ref := range api.Spec.SandHTTPRouteRefs {
-				if ref != "" {
-					httpRoutes = append(httpRoutes,
-						types.NamespacedName{
-							Namespace: api.Namespace,
-							Name:      ref,
-						}.String())
+			if api.Spec.Sandbox != nil {
+				for _, ref := range api.Spec.Sandbox[0].HTTPRouteRefs {
+					if ref != "" {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/controllers/dp/api_controller.go:928]: this is already validated in api webhook(isEmptyStringsInArray)

File: adapter/pkg/operator/controllers/dp/api_controller.go (hunk 2 above)
Comment (Krishanx92) [reply adapter/pkg/operator/controllers/dp/api_controller.go]: Fixed

File: adapter/pkg/operator/controllers/dp/api_controller.go (hunk 3 above)
Comment (Krishanx92) [reply adapter/pkg/operator/controllers/dp/api_controller.go]: Fixed

File: adapter/pkg/operator/controllers/dp/api_controller.go (hunk 4 above)
Comment (Krishanx92) [reply adapter/pkg/operator/controllers/dp/api_controller.go]: Fixed

File: adapter/pkg/operator/controllers/dp/api_controller.go (hunk 7 above)
Comment (Krishanx92) [reply adapter/pkg/operator/controllers/dp/api_controller.go:924]: This check need to add for resolve array index out exception

File: adapter/pkg/operator/apis/dp/v1alpha1/api_types.go [hunk 8]
```diff
@@ -87,6 +87,12 @@ type APISpec struct {
 	SystemAPI bool `json:"systemAPI"`
 }
 
+// EnvConfig contains the environment specific configuration
+type EnvConfig struct {
+	// Env denotes the environment of the API.
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/api_types.go:92]:
```suggestion
	// HTTPRouteRefs denotes the environment of the API.
```

File: adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go [hunk 9]
```diff
@@ -110,18 +110,26 @@ func (r *API) validateAPI() error {
 		r.Spec.APIType = "REST"
 	}
 
-	if !(r.Spec.ProdHTTPRouteRefs != nil && len(r.Spec.ProdHTTPRouteRefs) > 0) && !(r.Spec.SandHTTPRouteRefs != nil && len(r.Spec.SandHTTPRouteRefs) > 0) {
+	if !(r.Spec.Production != nil && r.Spec.Production[0].HTTPRouteRefs != nil && len(r.Spec.Production[0].HTTPRouteRefs) > 0) && !(r.Spec.Sandbox != nil && r.Spec.Sandbox[0].HTTPRouteRefs != nil && len(r.Spec.Sandbox[0].HTTPRouteRefs) > 0) {
```
Comment (AmaliMatharaarachchi) [inline adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go:113]: AFAIR, len would return 0 for nil obj as well. 
```suggestion
	if !(len(r.Spec.Production) > 0 && len(r.Spec.Production[0].HTTPRouteRefs) > 0) && !(len(r.Spec.Sandbox) > 0 && len(r.Spec.Sandbox[0].HTTPRouteRefs) > 0) {
```

File: adapter/pkg/operator/apis/dp/v1alpha1/api_types.go (hunk 8 above)
Comment (Krishanx92) [reply adapter/pkg/operator/apis/dp/v1alpha1/api_types.go]: Fixed

File: adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go (hunk 9 above)
Comment (Krishanx92) [reply adapter/pkg/operator/apis/dp/v1alpha1/api_webhook.go]: Fixed

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**